flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/UnionResultSubpartitionView.java

```diff
@@ -0,0 +1,248 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.partition;
+
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Queue;
+
+/**
+ * A wrapper to union the output from multiple {@link ResultSubpartitionView}s. This class provides
+ * the following guarantees to the output buffers.
+ *
+ * <ul>
+ *   <li>Each output buffer corresponds to a buffer in one of the subpartitions.
+ *   <li>Buffers in the same subpartition are output without their order changed.
+ *   <li>If a record is split and placed into multiple adjacent buffers due to the capacity limit of
+ *       the buffer, these buffers will be output consecutively without the entry of buffers from
+ *       other subpartitions in between.
+ * </ul>
+ */
+public class UnionResultSubpartitionView
+        implements ResultSubpartitionView, BufferAvailabilityListener {
+    private static final Logger LOGGER = LoggerFactory.getLogger(UnionResultSubpartitionView.class);
+
+    /** The maximum number of buffers to be cached in an instance of this class. */
```

Comment:
We'd better add more annotations to describe why it is necessary to cache some buffers in this view.